app/src/main/kotlin/compiler/intermediate/ControlFlow.kt

```diff
@@ -466,13 +526,77 @@ object ControlFlow {
 
                         is Statement.FunctionReturn -> {
                             addExpression(statement.value, functionReturnedValueVariables[Ref(function)])
+                            returning.addAll(last)
                             last = emptyList()
                         }
 
-                        is Statement.ForeachLoop -> TODO()
-                        is Statement.GeneratorYield -> TODO()
+                        is Statement.ForeachLoop -> {
+                            val gdg = getGeneratorDetailsGenerator(nameResolution[Ref(statement.generatorCall)]!!.value as Function)
+                            val frameMemoryAddress = if (function.isGenerator)
+                                getGeneratorDetailsGenerator(function).getNestedForeachFramePointerAddress(statement)
```

Comment:
If i understand correctly, we cannot use `getNestedForeachFramePointerAddress(statement)` there, because foreign generators don't have it implemented. I think we should check if the implementation of that generator is local, and if it's not, then `frameMemoryAddress=null` should be imo ok.

Reply:
Actually, this line is OK. `function` is given as a parameter in `processFunction` and thus have to be local.